agent/eni/iphelperwrapper/iphelper_windows.go

```diff
@@ -0,0 +1,155 @@
+	InitialNotificationTimeout = 4 * time.Second
+)
+
+var (
+	moduleiphlpapi = windows.NewLazySystemDLL("iphlpapi.dll")
+
+	procNotifyIPInterfaceChange = moduleiphlpapi.NewProc("NotifyIpInterfaceChange")
+	proccancelMibChangeNotify2  = moduleiphlpapi.NewProc("CancelMibChangeNotify2")
+
+	// Added these to allow unit testing for this module. We will swap these System calls with dummy calls in unit tests
+	funcNotifyIPInterfaceChange = procNotifyIPInterfaceChange.Call
+	funcCancelMibChangeNotify2  = proccancelMibChangeNotify2.Call
+
+	// The notification channel is used for communication between the notification callback from Windows APIs and the consumer
+	notificationChannel chan int
+)
+
+// Interface for using the monitor
+// We will mock this interface while unit testing other packages
+type InterfaceMonitor interface {
+	Start(notification chan int) error
+	Close() error
+}
+
+type eniMonitor struct {
+	notificationHandle windows.Handle
+}
+
+// Creates a new monitor to implement the interface
+func NewMonitor() (monitor *eniMonitor) {
+	monitor = &eniMonitor{}
+
+	notificationHandle := windows.Handle(0)
+	monitor.notificationHandle = notificationHandle
+
+	return
+}
+
+// This method is used to start the monitoring of new interfaces attached to the instance.
+// We register a notification callback with Windows API for any changes in instance interfaces (NotifyIPInterfaceChange).
+// Initial notification received within timeout is considered as a confirmation of successful registration
+// notificationChannel is used for any communication between the callback and the consumer
+func (monitor *eniMonitor) Start(notification chan int) error {
+	notificationChannel = notification
+
+	log.Info("IPHelper : Starting the eni monitor. Trying to register the callback from Windows API")
+	retVal, _, _ := funcNotifyIPInterfaceChange(syscall.AF_UNSPEC,
+		syscall.NewCallback(callback),
+		uintptr(unsafe.Pointer(&(monitor))),
+		1,
+		uintptr(unsafe.Pointer(&(monitor.notificationHandle))))
+
+	// If there is any error while registering callback then retVal will not be 0. Therefore, we will return with an error.
+	if retVal != 0 {
+		return errors.Errorf("error occured while calling Windows API : Error Msg : %s", syscall.Errno(retVal))
+	}
+
+	// Initial notification is sent immediately once the callback is registered.
+	// If the same is not received within a defined timeout then we assume an error and close the monitor
+	for {
+		select {
+		case event := <-notificationChannel:
+			if event == InitialNotification {
+				log.Info("IPHelper : NotifyIPInterfaceChange Windows API call successful. ENI Monitor has been started.")
+				return nil
+			}
+		case <-time.After(InitialNotificationTimeout):
+			log.Errorf("IPHelper : Initial notification not received within timeout. Closing the Monitor.")
+			if retErr := monitor.Close(); retErr != nil {
+				return errors.Wrapf(retErr, "initial notification not received as well")
```

Comment:
retErr would have the information about error while notification cancellation. It is warped along with the message "initial notification not received as well".
To avoid redundancy I had ignored this message.